### PR TITLE
[WIP] Issue 2889

### PR DESCRIPTION
**Pull request by werner291**
### Description

See the discussion in Issue #2889.

TL;DR: Bullet-based CCD is currently implemented by taking the pairwise convex hulls of each collision ibject in the robot model in the "before" and "after" states, and performing a collision test against that.

If the robot's movement is a pure translation (and assuming all collision objects are convex), this corresponds perfectly with the swept volume of the robot during this movement. However, if the movement includes rotation, especially with large swinging motions, this is a severe under-approximation.

What this PR contributes is the addition of sub-steps in the collision detection method depending on the amount of rotation in the motion. This has the advantage of allowing for very few (if any) sub-steps for linear or near-linear motions, but still providing a far better approximation if there is rotation in the movement.

I specifically made the following changes:
- Added `JointModel::distanceRotation` as a virtual method to get an idea of the rotation induced by a state change on a per-joint basis. Name is borrowed from `FloatingJointModel` for consistency; change should be non-breaking for that class.
- Added `RobotState::maximumRotation` which combines the per-joint rotations to compute an upper bound on every link's rotation.
- Made necessary changes to `CollisionEnvBullet::checkRobotCollisionHelperCCD` and added a parameter `max_rotation_per_step_` that controls the number of sub-steps.

PR is currently WIP since I'd like some feedback on whether this method works as I'd expect, and on how to best test it.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by werner291** [comment]
Sorry, I have no idea how I requested five different reviews.

**Comment by werner291** [comment]
Question: Why does the `RobotState::distance`` need the `JointModelGroup` exactly? And why not in the collision environment? What is the intended idea here?

**Comment by simonschmeisser** [comment]
> Sorry, I have no idea how I requested five different reviews.

By contributing to areas of code that have five different code owners registered ... don't worry that they will all respond though ;-)



> Question: Why does the `RobotState::distance` need the `JointModelGroup` exactly? And why not in the collision environment? What is the intended idea here?

It's the distance in joint space afaik, so you can limit that summation to only a certain subset of joints, e.g. only the gripper or only one arm

File: MIGRATION.md [hunk 1]
```diff
@@ -3,6 +3,10 @@
 API changes in MoveIt releases
 
 ## ROS Noetic
+- Added method `maximumRotation` to `RobotState`.
```
Comment (v4hn) [inline MIGRATION.md:6]: please leave out that line. only things relevant during migration are added here.

File: MIGRATION.md [hunk 2]
```diff
@@ -3,6 +3,10 @@
...
+- The method `distanceRotation` is now a virtual method of `JointModel` and implemented for all joint models.
+- Modified `CollisionEnvBullet`. It now takes a parameter `max_rotation_per_step_`, and subdivides CCD requests into sub-steps to better approximate rotations
```
Comment (v4hn) [inline MIGRATION.md:8]: This addition is API-compatible, so no need to mention it here either.

File: MIGRATION.md [hunk 3]
```diff
@@ -3,6 +3,10 @@
...
+- Renamed `addAttachedOjects` to `createAttachedCollisionObjects` in `CollisionEnvBullet` to reflect the non-mutating nature of the method.
```
Comment (v4hn) [inline MIGRATION.md:9]: I'm not sure this should be mentioned either. It's an API change, but it's not user-facing unless users hack in the deep internals for their own projects. Usually only `collision_detection` interfaces should be used.

File: moveit_core/collision_detection_bullet/src/collision_env_bullet.cpp [hunk 4]
```diff
@@ -208,27 +209,65 @@ void CollisionEnvBullet::checkRobotCollisionHelperCCD(const CollisionRequest& re
...
+    {
+      manager_CCD_->setCastCollisionObjectsTransform(link, st1.getCollisionBodyTransform(link, 0),
+                                                     st2.getCollisionBodyTransform(link, 0));
+    }
+
+    // Perform the contact test
+    manager_CCD_->contactTest(res, req, acm, false);
+
+    // Stop the operation after the first collision.
+    if (res.collision)
```
Comment (v4hn) [inline moveit_core/collision_detection_bullet/src/collision_env_bullet.cpp:266]: you should actually respect `req.max_contacts` here and keep going until you found enough contacts.
That will probably also fix your invalidated CI test because you miss collisions that only happen further into the trajectory.

To abort checking on collision, users can do checks for just one contact.

File: moveit_core/collision_detection_bullet/test/test_bullet_continuous_collision_checking.cpp [hunk 5]
```diff
@@ -368,6 +368,75 @@ TEST(ContinuousCollisionUnit, BulletCastMeshVsBox)
...
+
+  // Should be in collision when upright
+  cenv_->checkRobotCollision(req, res, upright, *acm_);
+  ASSERT_TRUE(res.collision);
+  res.clear();
+
+  // Now the CCD collision check should show the same as the previous check, as the robot hits the box when moving.
+  cenv_->checkRobotCollision(req, res, before_swing, after_swing, *acm_);
+  ASSERT_TRUE(res.collision);
+}
```
Comment (v4hn) [inline moveit_core/collision_detection_bullet/test/test_bullet_continuous_collision_checking.cpp:438]: I love this check. Thank you so much!

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 6]
```diff
@@ -1502,6 +1502,9 @@ class RobotState
   /** \brief Return the sum of joint distances to "other" state. Only considers active joints. */
   double distance(const RobotState& other, const JointModelGroup* joint_group) const;
 
+  /** \brief Return an upper bound on the rotation angle (radians) of any one link of the robot between two states. */
+  double maximumRotation(const RobotState& other) const;
+
```
Comment (v4hn) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h:1507]: This interface requires a lot more collision checks in your approach than are actually needed, because you can't consider the rotation distances separately for the different collision bodies. The upper bound of rotation for the base links will usually be drastically lower than for the end effector links.

Do you think it might be worth the effort to compute the distances for all collision bodies individually and discretize their path based on these values instead? The obvious downside would be the additional complexity of constructing possibly even more robot states and selectively checking only a subset of bodies in each one.
I'm asking because I work with robots with 50+ links and bimanual motion, so checking one arm for collisions over and over again just because the other one moves a lot sounds like a lot of overhead. That being said, I personally do not use bullet checking at all at the moment and my setup is surely not the average use case.

File: moveit_core/robot_state/src/robot_state.cpp [hunk 7]
```diff
@@ -920,6 +920,18 @@ double RobotState::distance(const RobotState& other, const JointModelGroup* join
...
 }
 
+double RobotState::maximumRotation(const RobotState& other) const
+{
+  double d = 0.0;
+  const std::vector<const JointModel*>& jm = this->getRobotModel()->getActiveJointModels();
+  for (const JointModel* joint : jm)
+  {
+    const int idx = joint->getFirstVariableIndex();
+    d += joint->distanceRotation(position_ + idx, other.position_ + idx);
```
Comment (v4hn) [inline moveit_core/robot_state/src/robot_state.cpp:930]:
```suggestion 
    d += joint->distanceRotation(&position_[idx], &other.position_[idx]);
```

It might be just me, but I consider this much more readable.
